### ISSUE TITLE
Wrong type for Categorical after write/load

### ISSUE THREAD

**Issue by shukon** [automl/ConfigSpace#100]
It is possible to define ints and bools for the choices of a Categorical. However, after saving to file, they are indistinguishable from strings. So on reloading, the configspaces are not the same anymore and configuration(-dict) from one configspace cannot be loaded by the other.

MWE:
```
import ConfigSpace as CS
import ConfigSpace.hyperparameters as CSH
from ConfigSpace.read_and_write import pcs_new
from ConfigSpace.configuration_space import Configuration

cs = CS.ConfigurationSpace()
cs.add_hyperparameter(CSH.CategoricalHyperparameter('int', [1, 2, 3]))
cs.add_hyperparameter(CSH.CategoricalHyperparameter('string', ['not', 'again']))
cs.add_hyperparameter(CSH.CategoricalHyperparameter('bool', [True, False]))

sample = cs.sample_configuration()

with open('configspace.pcs', 'w') as fh:
    fh.write(pcs_new.write(cs))
with open('configspace.pcs', 'r') as fh:
    cs_reloaded = pcs_new.read(fh.readlines())

Configuration(cs, sample.get_dictionary())           # Works!
Configuration(cs_reloaded, sample.get_dictionary())  # Doesn't
assert(cs_reloaded == cs)                            # Fails

# Passes, but should fail:
Configuration(cs_reloaded, {'int' : '3', 'string' : 'not', 'bool': 'True'})

# Fails, but shouldn't:
Configuration(cs_reloaded, {'int' : 3, 'string' : 'not', 'bool': True})
```

Related to #78 (because the error-message implies it to be the other way around, as we'd be trying to set a string where an int is needed. But in fact we try to (correctly) set an int, where a string is needed.) and maybe loosely to #73 and #25.

**Comment by mlindauer** [comment]
Hi,

Please try to write and load the configspace in the json format and not in the pcs format (see SMAC code for an example).
In the pcs format, we have no chance of rediscovering the types easily. The json format makes a lot of stuff much easier (at least for primitives).

Cheers,
Marius

**Comment by mfeurer** [comment]
Is this fixed by #148 @shukon ?

**Comment by shukon** [comment]
No, not really, since the `fix_types` only fixes `Configuration`s if the `ConfigurationSpace` object is given... but using the [json-format](https://github.com/automl/ConfigSpace/blob/master/ConfigSpace/read_and_write/json.py) does and I don't think the pcs-format should be used anymore anyway and I'd say it's ok that it's broken for special cases like the one mentioned above.

@mfeurer Anyway, maybe [this TODO](https://github.com/automl/ConfigSpace/blob/73ef3d30351a37c5c2505bfabf59e7f33d7aa514/ConfigSpace/hyperparameters.pyx#L1150) would help to reduce confusion about whether or not booleans are supported as Categoricals?

<details>
<summary>
Fixed code for anyone who needs it
</summary>
```
import ConfigSpace as CS
import ConfigSpace.hyperparameters as CSH 
from ConfigSpace.read_and_write import pcs_new, json
from ConfigSpace.configuration_space import Configuration

cs = CS.ConfigurationSpace()
cs.add_hyperparameter(CSH.CategoricalHyperparameter('int', [1, 2, 3]))
cs.add_hyperparameter(CSH.CategoricalHyperparameter('string', ['not', 'again']))
cs.add_hyperparameter(CSH.CategoricalHyperparameter('bool', [True, False]))

sample = cs.sample_configuration()

with open('configspace.pcs', 'w') as fh: 
    fh.write(json.write(cs))
with open('configspace.pcs', 'r') as fh: 
    cs_reloaded = json.read(fh.read())

Configuration(cs, sample.get_dictionary())           # Works!
Configuration(cs_reloaded, sample.get_dictionary())  # Works!
assert(cs_reloaded == cs)                            # Works!

# Fails and should fail:
try:
    Configuration(cs_reloaded, {'int' : '3', 'string' : 'not', 'bool': 'True'})
except ValueError:
    pass
else:
    raise ValueError("This should fail!")

# Doesn't fail and shouldn't:
Configuration(cs_reloaded, {'int' : 3, 'string' : 'not', 'bool': True})
```
</details>
